### PR TITLE
Firmware: Add multiple other options to move firmware to Linux

### PR DESCRIPTION
**Pull request by AdityaGarg8**

### CODE REVIEW

**Comment by AdityaGarg8** [comment]
@soopyc , any chance the changes in this PR can be useful for nixOS?

**Comment by AdityaGarg8** [comment]
This PR adds 2 new options:

1. Adds ability to rename firmware using python on macOS.
2. Adds ability to create distribution packages (currently deb has been added only).

I don't know how to make rpm packages, would need @sharpenedblade 's help

**Comment by soopyc** [comment]
On 5/12/24 02:00, Aditya Garg wrote:
> @soopyc , any chance the changes in this PR can be useful for nixOS?
> 

Option 2 looks great for use in exotic distros, though I've yet to test. 
Should still work though. Thanks!

-- 
Best regards,
Cassie Cheung (she/they)

File: docs/tools/firmware.sh
Comment (sharpenedblade) [inline docs/tools/firmware.sh:34]: Only ask people to install stuff if they dont already have it:

```
if ! command -v COMMAND &> /dev/null; then
    # ask to install it here, then return a error
fi
```

Also do this where you ask people to install python later in the script. You don't need to ask people to continue, just exit if the command is not available.

File: docs/tools/firmware.sh
Comment (sharpenedblade) [inline docs/tools/firmware.sh:36]: Indent this whole function

File: docs/tools/firmware.sh
Comment (sharpenedblade) [inline docs/tools/firmware.sh:143]: Run the `prerequisite_packages` function here instead, it does the same thing.

File: docs/tools/firmware.sh
Comment (sharpenedblade) [inline docs/tools/firmware.sh:168]: Make `prerequisite_packages` return a non-zero exit status if dependencies are not installed, then remove this code and replace it with `prerequisite_packages || exit 1`

File: docs/tools/firmware.sh
Comment (sharpenedblade) [inline docs/tools/firmware.sh:220]: Can you revert this (here and in the python script), I think its better if the tarball is the contents of `/usr/lib/firmware` so we can add other dirs into there if we need to.

**Review by sharpenedblade — CHANGES_REQUESTED** [review]

File: docs/tools/firmware.sh
Comment (AdityaGarg8) [reply docs/tools/firmware.sh:36]: As in add it to the macOS only part?

File: docs/tools/firmware.sh
Comment (AdityaGarg8) [reply docs/tools/firmware.sh:220]: I did the change so that users who choose option 2 don't get confused. I wouldn't be surprised if I get people copying the brcm folder in brcm of linux.

File: docs/tools/firmware.sh
Comment (sharpenedblade) [reply docs/tools/firmware.sh:36]: I mean just press tab in front of the contents of the create_deb function so its easy to see where it ends, but its not really that important

File: docs/tools/firmware.sh
Comment (sharpenedblade) [reply docs/tools/firmware.sh:220]: I guess its fine then

File: docs/tools/firmware.sh
Comment (sharpenedblade) [inline docs/tools/firmware.sh:236]: Word splitting can cause problems here, leave it in double quotes

**Review by sharpenedblade — CHANGES_REQUESTED** [review]
Everything looks good now, just change the `sudo python3` command

File: docs/tools/firmware.sh
Comment (AdityaGarg8) [reply docs/tools/firmware.sh:236]: My bad, that change was unintentional.

**Comment by AdityaGarg8** [comment]
Latest commit adds Arch package as well. @NoaHimesaka1873 @Redecorating , can you test the Arch bit whenever you get time?

File: docs/tools/firmware.sh
Comment (sharpenedblade) [inline docs/tools/firmware.sh:124]: Move the tar into `SOURCES` instead of unpacking it

**Review by sharpenedblade — APPROVED** [review]
Can you merge https://github.com/t2linux/wiki/pull/536, otherwise this looks good. If the other maintainers are fine with it then LGTM

**Comment by AdityaGarg8** [comment]
A final request for review before we can merge this. IMO its complete now.

**Comment by AdityaGarg8** [comment]
I just remembered, the wiki will also have to be updated:(

**Review by sharpenedblade — APPROVED** [review]
> A final request for review before we can merge this. IMO its complete now.

The code looks good to me. Lets wait for @Redecorating or @NoaHimesaka1873 to look at the arch code.

**Comment by NoaHimesaka1873** [comment]
> > A final request for review before we can merge this. IMO its complete now.
> 
> 
> 
> The code looks good to me. Lets wait for @Redecorating or @NoaHimesaka1873 to look at the arch code. 

@sharpenedblade @AdityaGarg8 Will test this in Wednesday when I have free time.

**Comment by AdityaGarg8** [comment]
> > > A final request for review before we can merge this. IMO its complete now.
> > 
> > 
> > The code looks good to me. Lets wait for @Redecorating or @NoaHimesaka1873 to look at the arch code.
> 
> @sharpenedblade @AdityaGarg8 Will test this in Wednesday when I have free time.

Thanks. Also, I'm not really sure whether modprobe -r brcmfmac && modprobe brcmfmac is needed on Arch as a post install script. I haven't added them yet for now.

**Comment by AdityaGarg8** [comment]
Updated the Wi-Fi guide as well

**Comment by AdityaGarg8** [comment]
https://adityagarg8.github.io/wiki/guides/wifi-bluetooth/

For a preview

**Comment by sharpenedblade** [comment]
> Thanks. Also, I'm not really sure whether modprobe -r brcmfmac && modprobe brcmfmac is needed on Arch as a post install script. I haven't added them yet for now.

If its not there on arch I would prefer that its also removed from the rpm package since it doesnt really belong there and telling users to reboot isnt that bad.

**Review by sharpenedblade — APPROVED** [review]
The doc changes look good

File: docs/guides/wifi-bluetooth.md
Comment (soopyc) [inline docs/guides/wifi-bluetooth.md:34]:
```suggestion
    **Method 1: Run the same script on Linux**
```

Maybe use bold fonts instead of headings so they don't show up in the sidebar while not being clickable. Same goes for the rest.

**Review by soopyc — COMMENTED** [review]
Looks fine apart from the headings.

**Comment by AdityaGarg8** [comment]
> > Thanks. Also, I'm not really sure whether modprobe -r brcmfmac && modprobe brcmfmac is needed on Arch as a post install script. I haven't added them yet for now.
> 
> If its not there on arch I would prefer that its also removed from the rpm package since it doesnt really belong there and telling users to reboot isnt that bad.

When I tried installing on a Live EOS ISO, it was trying to remake the initramfs inspite of not specifying it. But since the live CD doesn't have a vmlinuz in /boot, it was failing. So I thought on a real installation, Pacman maybe smart enough to reprobe the driver. I don't think there is anything wrong in simply running a modprobe -r brcmfmac after installing firmware. Why force a user to reboot when there is a simpler way. I guess we should try not to complicate things for users just for sake of *standardization* or *it works but doesn't belong here*.

**Comment by sharpenedblade** [comment]
> When I tried installing on a Live EOS ISO, it was trying to remake the initramfs inspite of not specifying it. But since the live CD doesn't have a vmlinuz in /boot, it was failing. So I thought on a real installation, Pacman maybe smart enough to reprobe the driver. I don't think there is anything wrong in simply running a modprobe -r brcmfmac after installing firmware. Why force a user to reboot when there is a simpler way. I guess we should try not to complicate things for users just for sake of standardization or it works but doesn't belong here.

AFAIK modprobe doesnt call any package manager stuff and just tells the kernel to load a module, but I guess its fine to leave it in, just add a note to the arch/eos section to reboot.

**Comment by AdityaGarg8** [comment]
> > When I tried installing on a Live EOS ISO, it was trying to remake the initramfs inspite of not specifying it. But since the live CD doesn't have a vmlinuz in /boot, it was failing. So I thought on a real installation, Pacman maybe smart enough to reprobe the driver. I don't think there is anything wrong in simply running a modprobe -r brcmfmac after installing firmware. Why force a user to reboot when there is a simpler way. I guess we should try not to complicate things for users just for sake of standardization or it works but doesn't belong here.
> 
> AFAIK modprobe doesnt call any package manager stuff and just tells the kernel to load a module, but I guess its fine to leave it in, just add a note to the arch/eos section to reboot.

I guess PKGBUILD has a post_install section as well. I'll probably add it later. Will also have to test on an Arch installation.

**Review by soopyc — APPROVED** [review]
Looks good from my end.

**Comment by NoaHimesaka1873** [comment]
> > > > A final request for review before we can merge this. IMO its complete now.
> > > 
> > > 
> > > The code looks good to me. Lets wait for @Redecorating or @NoaHimesaka1873 to look at the arch code.
> > 
> > 
> > @sharpenedblade @AdityaGarg8 Will test this in Wednesday when I have free time.
> 
> Thanks. Also, I'm not really sure whether modprobe -r brcmfmac && modprobe brcmfmac is needed on Arch as a post install script. I haven't added them yet for now.

@AdityaGarg8 It works. LGTM

**Review by NoaHimesaka1873 — APPROVED** [review]
LGTM

**Comment by NoaHimesaka1873** [comment]
> > > > > > A final request for review before we can merge this. IMO its complete now.
> > > > > 
> > > > > 
> > > > > The code looks good to me. Lets wait for @Redecorating or @NoaHimesaka1873 to look at the arch code.
> > > > 
> > > > 
> > > > @sharpenedblade @AdityaGarg8 Will test this in Wednesday when I have free time.
> > > 
> > > 
> > > Thanks. Also, I'm not really sure whether modprobe -r brcmfmac && modprobe brcmfmac is needed on Arch as a post install script. I haven't added them yet for now.
> > 
> > 
> > @AdityaGarg8 It works. LGTM
> 
> Is it reloading the driver automatically?

No. For that you have to create `apple-firmware.install`, put `post_install()` function there, and add `install=apple-firmware.install` to PKGBUILD.

**Comment by AdityaGarg8** [comment]
> > > > > > > A final request for review before we can merge this. IMO its complete now.
> > > > > > 
> > > > > > 
> > > > > > The code looks good to me. Lets wait for @Redecorating or @NoaHimesaka1873 to look at the arch code.
> > > > > 
> > > > > 
> > > > > @sharpenedblade @AdityaGarg8 Will test this in Wednesday when I have free time.
> > > > 
> > > > 
> > > > Thanks. Also, I'm not really sure whether modprobe -r brcmfmac && modprobe brcmfmac is needed on Arch as a post install script. I haven't added them yet for now.
> > > 
> > > 
> > > @AdityaGarg8 It works. LGTM
> > 
> > 
> > Is it reloading the driver automatically?
> 
> No. For that you have to create `apple-firmware.install`, put `post_install()` function there, and add `install=apple-firmware.install` to PKGBUILD. Reference: https://wiki.archlinux.org/title/PKGBUILD#install

Should apple-firmware.install be in the same directory as PKGBUILD?

**Comment by NoaHimesaka1873** [comment]
> > > > > > > > A final request for review before we can merge this. IMO its complete now.
> > > > > > > 
> > > > > > > 
> > > > > > > The code looks good to me. Lets wait for @Redecorating or @NoaHimesaka1873 to look at the arch code.
> > > > > > 
> > > > > > 
> > > > > > @sharpenedblade @AdityaGarg8 Will test this in Wednesday when I have free time.
> > > > > 
> > > > > 
> > > > > Thanks. Also, I'm not really sure whether modprobe -r brcmfmac && modprobe brcmfmac is needed on Arch as a post install script. I haven't added them yet for now.
> > > > 
> > > > 
> > > > @AdityaGarg8 It works. LGTM
> > > 
> > > 
> > > Is it reloading the driver automatically?
> > 
> > 
> > No. For that you have to create `apple-firmware.install`, put `post_install()` function there, and add `install=apple-firmware.install` to PKGBUILD. Reference: https://wiki.archlinux.org/title/PKGBUILD#install
> 
> Should apple-firmware.install be in the same directory as PKGBUILD?

Yes.

**Comment by AdityaGarg8** [comment]
Can you test now @NoaHimesaka1873 . I've pushed a change.

**Comment by AdityaGarg8** [comment]
> Can you test now @NoaHimesaka1873 . I've pushed a change.

It's regarding automatic reloading of driver btw

**Comment by AdityaGarg8** [comment]
> > > > > > > A final request for review before we can merge this. IMO its complete now.
> > > > > > 
> > > > > > 
> > > > > > The code looks good to me. Lets wait for @Redecorating or @NoaHimesaka1873 to look at the arch code.
> > > > > 
> > > > > 
> > > > > @sharpenedblade @AdityaGarg8 Will test this in Wednesday when I have free time.
> > > > 
> > > > 
> > > > Thanks. Also, I'm not really sure whether modprobe -r brcmfmac && modprobe brcmfmac is needed on Arch as a post install script. I haven't added them yet for now.
> > > 
> > > 
> > > @AdityaGarg8 It works. LGTM
> > 
> > 
> > Is it reloading the driver automatically?
> 
> No. For that you have to create `apple-firmware.install`, put `post_install()` function there, and add `install=apple-firmware.install` to PKGBUILD. Reference: https://wiki.archlinux.org/title/PKGBUILD#install

The latest commit (https://github.com/t2linux/wiki/pull/533/commits/4037f5e77a2367289944da7ccf96f3e2beb6bb05) seems to be reloading the driver on atlease EOS live environment on a VM. I would be nice if I could get a feed back from a bare metal installation.

**Comment by AdityaGarg8** [comment]
@soopyc , do you want to make changes to the nixOS guide alongwith this PR? I've noticed something regarding wifi over their.

**Comment by soopyc** [comment]
On 5/23/24 18:14, Aditya Garg wrote:
> @soopyc , do you want to make changes to the nixOS guide alongwith this PR? I've noticed something regarding wifi over their.
> 

Sure, I'll work on that now.
-- 
Best regards,
Cassie Cheung (she/they)

**Comment by AdityaGarg8** [comment]
Alright I am merging this, @soopyc you can open a new PR for nixos. If you want to add support in some way to the script, you can do that as well.

**Pull request merged**
